### PR TITLE
Feat: search useful addon version automatically

### LINKED ISSUE #4181: [Feature] Search useful addon version automatically.
**Is your feature request related to a problem? Please describe.**

Since when enable an addon, if not set version flag will enable the latest version of addon. But if the system version cannot meet the requirement will generate an error. This is bad user experience, os propose we can search the useful addon version automatically when the  latest version cannot meet version constraint.

**Describe the solution you'd like**

1. First, we cannot download every version of addon iteratively. This is too heavy, we must record the version constraint in `index.yaml`.
2. Check every addon version in descending order until find a matched version.
3. Tell user find a proper version :xxx, install it [Y/N]. User should make the decision whether install it.
4. Only do this search when latest version cannot meet version constraint and not set the `--version` flag.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Hanmengnan**
Verify whether the current addon version meets the system version requirements according to the obtained specified version. There are two system version requirements: Vela core version, K8s version.

If meet the requirements and continue to perform the next task.

If the requirements are not met, obtain the highest version that meets the requirements

Refs  #4181

Signed-off-by: HanMengnan <1448189829@qq.com>

### Description of your changes

Fixes  #4181

The version verification logic can be designed as:

- Verify whether the current addon version meets the system version requirements according to the obtained specified version. There are two system version requirements: Vela core version, K8s version. The requirements can be describe by the meta.yaml file's annotations filed:
```yaml
apiVersion: v1
entries:
  fluxcd:
  - apiVersion: v2
    appVersion: 1.16.0
    description: A Helm chart for Kubernetes
    name: fluxcd
    type: application
    urls:
    - fluxcd-1.0.0.tgz
    version: 1.0.0
    annotations:
       system: "vela>=1.3.0; kubernetes>=1.10.0"
  - apiVersion: v2
    appVersion: 1.16.0
    description: A Helm chart for Kubernetes
    name: fluxcd
    type: application
    urls:
    - fluxcd-1.1.0.tgz
    version: 1.1.0
    annotations:
       system: "vela>=1.4.0; kubernetes>=1.20.0"
generated: "2022-06-15T13:17:04.733573+08:00"
```
 - If meet the requirements and continue to perform the next task.
 - If the requirements are not met, obtain the highest version that meets the requirements.
    - If can get the addon version that meets the requirements and ask the user whether to deploy with this version.
    - If the user agrees to deploy using the prompt version, deploy it.
- If the user disagrees, exit.
- Unable to get the addon version that meets the requirements. An error message is returned.

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Unit test and E2E test.
In the E2E test, I built a helm test repository in which there are two versions of a `fluxcd-test-version` image. The first version does not meet the system requirements, and the second version meets the requirements.

```yaml
apiVersion: v1
entries:
  fluxcd-test-version:
    - apiVersion: v2
      appVersion: 1.16.0
      description: A Helm chart for Kubernetes
      name: fluxcd-test-version
      type: application
      urls:
        - http://127.0.0.1:9098/helm/fluxcd-test-version-1.0.0.tgz
      version: 1.0.0
      annotations:
        system: "vela>=1.3.0; kubernetes>=1.10.0"
    - apiVersion: v2
      appVersion: 1.16.0
      description: A Helm chart for Kubernetes
      name: fluxcd-test-version
      type: application
      urls:
        - http://127.0.0.1:9098/helm/fluxcd-test-version-2.0.0.tgz
      version: 2.0.0
      annotations:
        system: "vela>=1.5.0; kubernetes>=1.30.0"
generated: "2022-06-15T13:17:04.733573+08:00"
```

### CODE REVIEW

**Comment by wangyikewxgm** [comment]
good joob! [catalog CI](https://github.com/kubevela/catalog/blob/44a2c070b29ea5a576f947bba8d23fcf66ce59e6/hack/addons/syn_addon_package.go#L106) need this change?

File: e2e/addon/addon_test.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ import (
 
 	v1 "k8s.io/api/core/v1"
 
+	"github.com/Netflix/go-expect"
```
Comment (wonderflow) [inline e2e/addon/addon_test.go:27]: why we need this package?

Comment (Hanmengnan) [reply e2e/addon/addon_test.go:27]: When implementing the E2E test, I used the `sendline` function of this package to simulate user input


```go
It("Enable fluxcd-test-version whose version can't suit system requirements", func() {
	output, err := e2e.InteractiveExec("vela addon enable fluxcd-test-version", func(c *expect.Console) {
		_, err = c.SendLine("y")
		Expect(err).NotTo(HaveOccurred())
	})
	Expect(output).To(ContainSubstring("enabled successfully"))
	Expect(err).NotTo(HaveOccurred())
})
```

I find that in `e2e/commoncontext.go` uses this method, so I choose to use the same function.

Comment (wonderflow) [reply e2e/addon/addon_test.go:27]: ok

File: pkg/addon/addon.go [hunk 2]
```diff
@@ -1393,6 +1394,26 @@ func (h *Installer) continueOrRestartWorkflow() error {
...
+			Password: h.r.Helm.Password,
+		})
+		versions, err := versionedRegistry.GetAddonAvailableVersion(addonName)
+		if err != nil {
+			return ""
+		}
+		for _, version := range versions {
+			req := LoadSystemRequirements(version.Annotations["system"])
+			if checkAddonVersionMeetRequired(h.ctx, req, h.cli, h.dc) == nil {
+				return version.Version
```
Comment (StevenLeiZhang) [inline pkg/addon/addon.go:1410]: If multiple addon versions match requirement. Should this function returns all of them, or just return one.
If it only returns one version, I think it should return the latest one.

**Comment by Hanmengnan** [comment]
> good joob! [catalog CI](https://github.com/kubevela/catalog/blob/44a2c070b29ea5a576f947bba8d23fcf66ce59e6/hack/addons/syn_addon_package.go#L106) need this change?

Yes, The `meta.systemrequirements` information needs to be loaded. I will submit a PR to modify the corresponding code.

File: pkg/addon/addon.go (hunk 2 above)
Comment (Hanmengnan) [reply pkg/addon/addon.go:1410]: Yes, getAddonVersionMeetSystemRequirement return the addon's latest version which meet the system requirements

**Review by StevenLeiZhang — APPROVED** [review]
LGTM

File: e2e/addon/mock/vela_addon_mock_server.go [hunk 3]
```diff
@@ -103,6 +105,29 @@ var ossHandler http.HandlerFunc = func(rw http.ResponseWriter, req *http.Request
...
+	case strings.Contains(req.URL.Path, "fluxcd-test-version-1.0.0.tgz"):
+		files, err := ioutil.ReadFile("./e2e/addon/mock/testrepo/helm-repo/fluxcd-test-version-1.0.0.tgz")
+		if err != nil {
+			_, _ = rw.Write([]byte(err.Error()))
+		}
+		rw.Write(files)
+	case strings.Contains(req.URL.Path, "fluxcd-test-version-2.0.0.tgz"):
+		files, err := ioutil.ReadFile("./e2e/addon/mock/testrepo/helm-repo/fluxcd-test-version-2.0.0.tgz")
+		if err != nil {
+			_, _ = rw.Write([]byte(err.Error()))
```
[the remark this answers is not in the document]
Comment (Hanmengnan) [reply e2e/addon/mock/vela_addon_mock_server.go:125]: Here, I read the file by imitating the method of the original test case. The content prompted by the robot is to use `template.execute()`method renders the contents of the file, which does not meet the requirements

File: e2e/addon/mock/vela_addon_mock_server.go [hunk 4]
```diff
@@ -103,6 +105,29 @@ var ossHandler http.HandlerFunc = func(rw http.ResponseWriter, req *http.Request
...
+	case strings.Contains(req.URL.Path, "index.yaml"):
+		files, err := ioutil.ReadFile("./e2e/addon/mock/testrepo/helm-repo/index.yaml")
+		if err != nil {
+			_, _ = rw.Write([]byte(err.Error()))
+		}
+		rw.Write(files)
+	case strings.Contains(req.URL.Path, "fluxcd-test-version-1.0.0.tgz"):
+		files, err := ioutil.ReadFile("./e2e/addon/mock/testrepo/helm-repo/fluxcd-test-version-1.0.0.tgz")
+		if err != nil {
+			_, _ = rw.Write([]byte(err.Error()))
```
[the remark this answers is not in the document]
Comment (Hanmengnan) [reply e2e/addon/mock/vela_addon_mock_server.go:119]: Here, I read the file by imitating the method of the original test case. The content prompted by the robot is to use `template.execute()` method renders the contents of the file, which does not meet the requirements

File: e2e/addon/mock/vela_addon_mock_server.go [hunk 5]
```diff
@@ -103,6 +105,29 @@ var ossHandler http.HandlerFunc = func(rw http.ResponseWriter, req *http.Request
...
+		}
+		rw.Write(files)
+	case strings.Contains(req.URL.Path, "fluxcd-test-version-2.0.0.tgz"):
+		files, err := ioutil.ReadFile("./e2e/addon/mock/testrepo/helm-repo/fluxcd-test-version-2.0.0.tgz")
+		if err != nil {
+			_, _ = rw.Write([]byte(err.Error()))
+		}
+		rw.Write(files)
```
[the remark this answers is not in the document]
Comment (Hanmengnan) [reply e2e/addon/mock/vela_addon_mock_server.go]: Here, I read the file by imitating the method of the original test case. The content prompted by the robot is to use `template.execute()` method renders the contents of the file, which does not meet the requirements

File: e2e/addon/mock/vela_addon_mock_server.go [hunk 6]
```diff
@@ -103,6 +105,29 @@ var ossHandler http.HandlerFunc = func(rw http.ResponseWriter, req *http.Request
...
+	case strings.Contains(req.URL.Path, "fluxcd-test-version-1.0.0.tgz"):
+		files, err := ioutil.ReadFile("./e2e/addon/mock/testrepo/helm-repo/fluxcd-test-version-1.0.0.tgz")
+		if err != nil {
+			_, _ = rw.Write([]byte(err.Error()))
+		}
+		rw.Write(files)
```
[the remark this answers is not in the document]
Comment (Hanmengnan) [reply e2e/addon/mock/vela_addon_mock_server.go]: Here, I read the file by imitating the method of the original test case. The content prompted by the robot is to use `template.execute()` method renders the contents of the file, which does not meet the requirements

File: e2e/addon/mock/vela_addon_mock_server.go [hunk 7]
```diff
@@ -103,6 +105,29 @@ var ossHandler http.HandlerFunc = func(rw http.ResponseWriter, req *http.Request
...
 }
 
+var helmHandler http.HandlerFunc = func(rw http.ResponseWriter, req *http.Request) {
+	switch {
+	case strings.Contains(req.URL.Path, "index.yaml"):
+		files, err := ioutil.ReadFile("./e2e/addon/mock/testrepo/helm-repo/index.yaml")
+		if err != nil {
+			_, _ = rw.Write([]byte(err.Error()))
+		}
+		rw.Write(files)
```
[the remark this answers is not in the document]
Comment (Hanmengnan) [reply e2e/addon/mock/vela_addon_mock_server.go]: Here, I read the file by imitating the method of the original test case. The content prompted by the robot is to use `template.execute()` method renders the contents of the file, which does not meet the requirements

File: e2e/addon/mock/vela_addon_mock_server.go [hunk 8]
```diff
@@ -103,6 +105,29 @@ var ossHandler http.HandlerFunc = func(rw http.ResponseWriter, req *http.Request
 	}
 }
 
+var helmHandler http.HandlerFunc = func(rw http.ResponseWriter, req *http.Request) {
+	switch {
+	case strings.Contains(req.URL.Path, "index.yaml"):
+		files, err := ioutil.ReadFile("./e2e/addon/mock/testrepo/helm-repo/index.yaml")
+		if err != nil {
+			_, _ = rw.Write([]byte(err.Error()))
```
[the remark this answers is not in the document]
Comment (Hanmengnan) [reply e2e/addon/mock/vela_addon_mock_server.go:113]: Here, I read the file by imitating the method of the original test case. The content prompted by the robot is to use `template.execute()` method renders the contents of the file, which does not meet the requirements

File: pkg/addon/addon.go [hunk 9]
```diff
@@ -1144,7 +1144,8 @@ func (h *Installer) enableAddon(addon *InstallPackage) error {
 	h.addon = addon
 	err = checkAddonVersionMeetRequired(h.ctx, addon.SystemRequirements, h.cli, h.dc)
 	if err != nil {
-		return VersionUnMatchError{addonName: addon.Name, err: err}
+		version := h.getAddonVersionMeetSystemRequirement(addon.Name)
+		return VersionUnMatchError{addonName: addon.Name, err: err, userSelectedAddonVersion: addon.Version, availableVersion: version}
 	}
 
 	if err = h.installDependency(addon); err != nil {
```
[the remark this answers is not in the document]
Comment (Hanmengnan) [reply pkg/addon/addon.go]: I didn't modify the `installDependency` function. It has always been so

File: references/cli/addon.go [hunk 10]
```diff
@@ -518,10 +514,21 @@ func enableAddon(ctx context.Context, k8sClient client.Client, dc *discovery.Dis
...
+				err = pkgaddon.EnableAddon(ctx, name, availableVersion, k8sClient, dc, apply.NewAPIApplicator(k8sClient), config, registry, args, nil)
+				return err
 			}
+			// The user does not agree to use the version provided by us
+			return fmt.Errorf("you can try another version by command: \"vela addon enable %s --version <version> \" ", name)
+		}
+		if err != nil {
 			return err
 		}
 		if err = waitApplicationRunning(k8sClient, name); err != nil {
```
[the remark this answers is not in the document]
Comment (Hanmengnan) [reply references/cli/addon.go:534]: I didn't modify the `waitApplicationRunning` function. It has always been so

File: pkg/addon/versioned_registry.go [hunk 11]
```diff
@@ -212,3 +231,22 @@ func chooseVersion(specifiedVersion string, versions []*repo.ChartVersion) (*rep
 	}
 	return addonVersion, availableVersions
 }
+
+// LoadSystemRequirements load the system version requirements from the addon's meta file
+func LoadSystemRequirements(requirements string) *SystemRequirements {
+	if len(requirements) == 0 {
+		return nil
+	}
+	patternReq := `vela(.*\d+\.\d+\.\d+);\s?kubernetes(.*\d+\.\d+\.\d+)`
```
Comment (wangyikewxgm) [inline pkg/addon/versioned_registry.go:240]: please define this string as a const str.

Comment (Hanmengnan) [reply pkg/addon/versioned_registry.go:240]: ok

**Review by wangyikewxgm — APPROVED** [review]
LGTM！Geate job！

**Review by wonderflow — APPROVED** [review]

**Pull request merged**